google/cloud/logging_v2/_instrumentation.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Add diagnostic instrumentation source information to logs"""
+import collections
+
+from google.cloud.logging_v2.entries import StructEntry
+from google.cloud.logging_v2 import __version__
+
+_DIAGNOSTIC_INFO_KEY = "logging.googleapis.com/diagnostic"
+_INSTRUMENTATION_SOURCE_KEY = "instrumentation_source"
+_PYTHON_LIBRARY_NAME = "python"
+
+_LIBRARY_VERSION = __version__
+
+_MAX_NAME_LENGTH = 14
+_MAX_VERSION_LENGTH = 14
+_MAX_INSTRUMENTATION_ENTRIES = 3
+
+
+def _add_instrumentation(entries, **kw):
+    """Validate or add instrumentation information to a list of entries
+
+        Ensures that a single log entry with valid instrumentation info
+        is in `entries`.  If instrumentation info was manually added,
+        it is validated to ensure it matches this library's information.
+        Otherwise, a new diagnostic entry is prepended to the list of
+        entries.
+
+    Args:
+       entries (Sequence[Mapping[str, ...]]): sequence of mappings representing
+            the log entry resources to log.
+
+    Returns:
+        Sequence[Mapping[str, ...]]: entries with instrumentation info validated if present
+        otherwise added to beginning of list.
+    """
+    is_written = False
+    new_entries = []
+    for entry in entries:
```

Comment:
Simplify by assuming we need to emit exactly once.

Reply:
On second thought, leave in for future expansion with different layers from handlers, etc...